### PR TITLE
[go] Generate inline enum models and prefix all enums

### PR DESCRIPTION
**Pull request by fantavlik**
### PR checklist

- [ ✓ ] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ ✓ ] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [ ✓ ] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`~~, `3.4.x`, `4.0.x`~~. *Default: `master`*.
- [ ✓ ] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

CC: @antihax @bvwells @grokify @kemokemo 

This aims to address the following issues:
https://github.com/OpenAPITools/openapi-generator/issues/535
(also see related closed PR: https://github.com/OpenAPITools/openapi-generator/pull/542)
https://github.com/OpenAPITools/openapi-generator/issues/2050
https://github.com/OpenAPITools/openapi-generator/issues/2360

### Description of the PR

There are four main aspects:
1. Enums that are defined inline were not being generated (see https://github.com/OpenAPITools/openapi-generator/issues/2360 for an example), this PR follows similar logic to the typescript-fetch generation template to remedy this: https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/typescript-fetch/modelGeneric.mustache#L92-L109
2. Enums names often result in namespace collisions within the package being generated (see https://github.com/OpenAPITools/openapi-generator/issues/535), prefixes have been added to help prevent this for most cases.
3. Numeric enum values were improperly being quoted, now only strings are.
4. Numeric enum names were not properly escaping plus/minus signs and decimal points e.g. to differentiate between `-1` and `1` or `1.1` and `11`

### CODE REVIEW

File: samples/client/petstore/go/go-petstore/model_enum_arrays.go [hunk 1]
```diff
@@ -10,6 +10,22 @@
 package petstore
 
 type EnumArrays struct {
-	JustSymbol string `json:"just_symbol,omitempty"`
-	ArrayEnum []string `json:"array_enum,omitempty"`
+	JustSymbol EnumArraysJustSymbol `json:"just_symbol,omitempty"`
+	ArrayEnum EnumArraysArrayEnum `json:"array_enum,omitempty"`
```
Comment (fantavlik) [inline samples/client/petstore/go/go-petstore/model_enum_arrays.go]: This isn't quite right, it needs to be
```go
ArrayEnum []EnumArraysArrayEnum```
on this line and
```go
type EnumArraysArrayEnum string```
one line 25 below

I'll work on a fix...

File: samples/client/petstore/go/go-petstore/model_map_test_.go [hunk 2]
```diff
@@ -11,7 +11,15 @@ package petstore
 
 type MapTest struct {
 	MapMapOfString map[string]map[string]string `json:"map_map_of_string,omitempty"`
-	MapOfEnumString map[string]string `json:"map_of_enum_string,omitempty"`
+	MapOfEnumString MapTestMapOfEnumString `json:"map_of_enum_string,omitempty"`
```
Comment (fantavlik) [inline samples/client/petstore/go/go-petstore/model_map_test_.go]: This isn't quite right, I think it needs to be
```go
ArrayEnum map[string] MapTestMapOfEnumString```
on this line and
```go
type MapTestMapOfEnumString string```
one line 19 below

I'll work on a fix...

File: samples/client/petstore/go/go-petstore/model_enum_arrays.go (hunk 1 above)
Comment (fantavlik) [reply samples/client/petstore/go/go-petstore/model_enum_arrays.go]: Fixed

File: samples/client/petstore/go/go-petstore/model_map_test_.go (hunk 2 above)
Comment (fantavlik) [reply samples/client/petstore/go/go-petstore/model_map_test_.go]: Fixed

**Comment by fantavlik** [comment]
Hello Go working group 👋  @antihax @bvwells @grokify @kemokemo I believe this is ready for review whenever you get a chance. There are some test failures related to adding types for the inline enums that were missing - I'm happy to spend time fixing those if you think this PR is heading in the right direction.

**Comment by wing328** [comment]
The CircleCI job failed with the following errors:
```
=== RUN   TestGetPetById
--- FAIL: TestGetPetById (0.00s)
	pet_api_test.go:299: 
			Error Trace:	pet_api_test.go:299
			            				pet_api_test.go:56
			Error:      	Not equal: 
			            	expected: petstore.PetStatus("pending")
			            	actual  : string("pending")
			Test:       	TestGetPetById
			Messages:   	Pet status should be pending
=== RUN   TestGetPetByIdWithInvalidID
--- PASS: TestGetPetByIdWithInvalidID (0.00s)
=== RUN   TestUpdatePetWithForm
--- FAIL: TestUpdatePetWithForm (0.01s)
	pet_api_test.go:299: 
			Error Trace:	pet_api_test.go:299
			            				pet_api_test.go:92
			Error:      	Not equal: 
			            	expected: petstore.PetStatus("available")
			            	actual  : string("available")
			Test:       	TestUpdatePetWithForm
			Messages:   	Pet status should be available
=== RUN   TestFindPetsByTag
--- FAIL: TestFindPetsByTag (0.00s)
	pet_api_test.go:110: 
			Error Trace:	pet_api_test.go:110
			Error:      	Not equal: 
			            	expected: petstore.PetStatus("available")
			            	actual  : string("available")
			Test:       	TestFindPetsByTag
			Messages:   	Pet status should be `pending`
=== RUN   TestFindPetsByStatus
--- FAIL: TestFindPetsByStatus (0.00s)
	pet_api_test.go:138: 
			Error Trace:	pet_api_test.go:138
			Error:      	Not equal: 
			            	expected: petstore.PetStatus("available")
			            	actual  : string("available")
			Test:       	TestFindPetsByStatus
			Messages:   	Pet status should be `available`
	pet_api_test.go:138: 
			Error Trace:	pet_api_test.go:138
			Error:      	Not equal: 
			            	expected: petstore.PetStatus("available")
			            	actual  : string("available")
			Test:       	TestFindPetsByStatus
			Messages:   	Pet status should be `available`
	pet_api_test.go:138: 
			Error Trace:	pet_api_test.go:138
			Error:      	Not equal: 
			            	expected: petstore.PetStatus("available")
			            	actual  : string("available")
			Test:       	TestFindPetsByStatus
			Messages:   	Pet status should be `available`
	pet_api_test.go:138: 
			Error Trace:	pet_api_test.go:138
			Error:      	Not equal: 
			            	expected: petstore.PetStatus("available")
			            	actual  : string("available")
			Test:       	TestFindPetsByStatus
			Messages:   	Pet status should be `available`
	pet_api_test.go:138: 
			Error Trace:	pet_api_test.go:138
			Error:      	Not equal: 
			            	expected: petstore.PetStatus("available")
			            	actual  : string("available")
			Test:       	TestFindPetsByStatus
			Messages:   	Pet status should be `available`
	pet_api_test.go:138: 
			Error Trace:	pet_api_test.go:138
			Error:      	Not equal: 
			            	expected: petstore.PetStatus("available")
			            	actual  : string("available")
			Test:       	TestFindPetsByStatus
			Messages:   	Pet status should be `available`
	pet_api_test.go:138: 
			Error Trace:	pet_api_test.go:138
			Error:      	Not equal: 
			            	expected: petstore.PetStatus("available")
			            	actual  : string("available")
			Test:       	TestFindPetsByStatus
			Messages:   	Pet status should be `available`
	pet_api_test.go:138: 
			Error Trace:	pet_api_test.go:138
			Error:      	Not equal: 
			            	expected: petstore.PetStatus("available")
			            	actual  : string("available")
			Test:       	TestFindPetsByStatus
			Messages:   	Pet status should be `available`
```

**Comment by fantavlik** [comment]
Hey @wing328 sorry for the late reply, I've been rethinking this approach.

This PR is actually a workaround for a more central issue. I think the correct approach would be to extend https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/InlineModelResolver.java to also flatten inline enums (and probably also inline `oneOf`, `allOf`, and `anyOf`) in addition to the currently supported flattening of objects and arrays of objects.

**Comment by wing328** [comment]
>  https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/InlineModelResolver.java to also flatten inline enums 

My take is that if the user wants to have the enum reusable throughout the spec, they should define it separately and use $ref instead. Is that what you mean or you're suggesting something else?

**Comment by fantavlik** [comment]
@wing328 the problem I'm hoping to solve is that many inline schemas are not being generated at all without [some really complex templating](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/typescript-fetch/modelGeneric.mustache#L92-L109). I agree that the spec author probably doesn't need these models to be reusable if they are defined inline but I think that's the only way reuse the existing *Generator.java model handing and existing model templates without having to worry about special-case handling for really complex inline models which may need to spawn many new models within a single file to handle properly - templates just don't seem to be setup to handle this case very well at the moment.

Consider this:
```yaml
components:
  schemas:
    MySchema:
      allOf:
      - $ref: '#/components/schemas/Foo'
      - type: object
        title: MyModel
        properties:
          thing:
            title: MyInnerThing
            type: object
            properties:
              anotherModel:
                title: MyStringEnum
                type: string
                enum: [on,off]
              yetAnotherModel:
                title: MyIntEnum
                type: integer
                enum: [0,2,4]
              andAnother:
                title: MyOneOf
                oneOf:
                  - $ref: '#/components/schemas/Baz'
                  - $ref: '#/components/schemas/Bar'
    Foo:
      type: object
      properties:
        fooProp:
          type: string
      required:
      - fooProp
    Bar:
      type: object
      properties:
        kind:
          type: string
        barProp:
          type: string
      required:
      - barProp
    Baz:
      type: object
      properties:
        kind:
          type: string
        bazProp:
          type: string
      required:
      - bazProp
```

We're currently generating:
model_my_schema.go
```go
package openapi

type MySchema struct {
	FooProp string `json:"fooProp"`
	Thing map[string]interface{} `json:"thing,omitempty"`
}
```

I think with some better handling of InlineModelResolver.java we would properly generate the missing models in their own files: `MyModel`, `MyInnerThing`, `MyStringEnum`, `MyIntEnum`, `MyOneOf`

Trying to a handle a case like this in a single model file using templates seems like a losing battle. What do you think?

**Comment by wild-endeavor** [comment]
Hey all - do you think this will get merged eventually?  We'd like to use it in our code generation bit.  Thanks.

**Comment by wild-endeavor** [comment]
Still no update on this?

**Comment by fantavlik** [comment]
@wild-endeavor I'm working on the core fix for all languages, I had other work take precedence. I'm not going to fix the tests for this implementation so I will close this PR and put up a separate one that's not specific to Go.

**Pull request closed without merging**